tokio/src/sync/mpsc/chan.rs

```diff
@@ -129,6 +130,49 @@ impl<T, S> Tx<T, S> {
         Tx { inner: chan }
     }
 
+    pub(super) fn downgrade(self) -> Self {
+        if self.inner.tx_count.fetch_sub(1, AcqRel) == 1 {
+            // Close the list, which sends a `Close` message
+            self.inner.tx.close();
+
+            // Notify the receiver
+            self.wake_rx();
+        }
+
+        self
+    }
+
+    // Returns a boolean that indicates whether the channel is closed.
+    pub(super) fn upgrade(self) -> Option<Self> {
+        let mut tx_count = self.inner.tx_count.load(Acquire);
+
+        if tx_count == 0 {
+            // channel is closed
+            return None;
+        }
```

Comment:
You have this `if` twice in this function. If you move it to the top of the loop, then you can remove the duplication.